optuna/distributions.py

```diff
@@ -233,6 +316,114 @@ def _contains(self, param_value_in_internal_repr: float) -> bool:
         return self.low <= value <= self.high and abs(k - round(k)) < 1.0e-8
 
 
+class IntDistribution(BaseDistribution):
+    """A discrete uniform distribution on integers.
+
+    This object is instantiated by :func:`~optuna.trial.Trial.suggest_float`, and passed to
+    :mod:`~optuna.samplers` in general.
+
+    Attributes:
+        low:
+            Lower endpoint of the range of the distribution. ``low`` is included in the range.
+        high:
+            Upper endpoint of the range of the distribution. ``high`` is included from the range.
+        log:
+            If ``log`` is True, this distribution is in log-scaled domain.
+        step:
+            A step for spacing between values. When ``log`` is True, ``step`` is assumed to be 1.
+
+    Raises:
+        ValueError:
+            If ``low`` value is larger than ``high`` value,
+            or ``low`` value is less than 1 when ``log`` is :obj:`True`.
+            If ``step`` is not positive value.
+    """
+
+    def __init__(self, low: int, high: int, log: Union[None, bool] = False, step: int = 1) -> None:
+        if low > high:
+            raise ValueError(
+                "The `low` value must be smaller than or equal to the `high` value "
+                "(low={}, high={}).".format(low, high)
+            )
+
+        # For backward compatibility, an abbreviated json for IntUniformDistribution
+        # may not have `log` attribute.
+        if log is None:
+            log = False
+
+        if log and low < 1:
+            raise ValueError(
+                "The `low` value must be equal to or greater than 1 for a log distribution "
+                "(low={}, high={}).".format(low, high)
+            )
+
+        if step <= 0:
+            raise ValueError(
+                "The `step` value must be non-zero positive value, but step={}.".format(step)
+            )
+
+        if log and step != 1:
+            self._warn_step()
```

Comment:
Since we will create a new `IntDistribution` class, we can raise `ValueError` here with the comment `"Samplers and other components in Optuna only accept step is 1 when log argument is True"`.

Also, in this case, we can remove some adhoc code such as

- _warn_step
- _asdict